Writing-editing network/seq2seq/discriminator.py

```diff
@@ -0,0 +1,129 @@
+import torch.nn as nn
+import torch
+import torch.nn.functional as F
+import numpy as np
+
+class Encoder(nn.Module):
+    def __init__(self, embedding_dim, hidden_dim, vocab_size, batch_size):
+        super(Encoder, self).__init__()
+        self.hidden_dim = hidden_dim
+        self.word_embeddings = nn.Embedding(vocab_size, embedding_dim)
+        self.lstm = nn.LSTM(embedding_dim, hidden_dim, batch_first=True)
+        self.batch_size = batch_size
+        self.hidden = self.init_hidden()
+
+    def init_hidden(self):
+        return (torch.zeros(1, self.batch_size, self.hidden_dim), torch.zeros(1, self.batch_size, self.hidden_dim))
+
+    def forward(self, abstract):
+        embeds = self.word_embeddings(abstract)
+        lstm_out, self.hidden = self.lstm(embeds.view(self.batch_size, len(abstract[0]), -1), self.hidden)
+        return lstm_out, self.hidden
+
+class DecoderRNN(nn.Module):
+    def __init__(self, embedding_size, hidden_size, vocab_size, output_size, batch_size):
+        super(DecoderRNN, self).__init__()
+        self.hidden_size = hidden_size
+        self.embedding_size = embedding_size
+
+        self.embedding = nn.Embedding(vocab_size, embedding_size)
+        self.lstm = nn.LSTM(embedding_size, hidden_size, batch_first=True)
+        self.out = nn.Linear(hidden_size, output_size)
+        self.batch_size = batch_size
+
+    def forward(self, input, hidden):
+        output = self.embeddingF(input).view(self.batch_size, 1, -1)
+        output = F.relu(output)
+        output, hidden = self.lstm(output, hidden)
+        output = self.out((output.view(self.batch_size, 1, -1)))
+        output = output.squeeze(2)
+        return output, hidden
+
+def decoder_train(input_tensor, encoder_hidden, decoder, seq_length):
+    decoder_input = torch.zeros((2,1), dtype=torch.long)
+    decoder_hidden = encoder_hidden
+    decoder_output_f = torch.tensor([])
+    for di in range(seq_length):
+        decoder_output, decoder_hidden = decoder(decoder_input, decoder_hidden)
+        decoder_input = input_tensor[:, di]
+        decoder_output_f = torch.cat((decoder_output_f, decoder_output), dim=1)
+    return decoder_output_f
+
+# critic
+class Critic(nn.Module):
+    def __init__(self, embedding_dim, hidden_dim, vocab_size):
+        super(Critic, self).__init__()
+        self.hidden_dim = hidden_dim
+        self.word_embeddings = nn.Embedding(vocab_size, embedding_dim)
+        self.lstm = nn.LSTM(embedding_dim, hidden_dim, batch_first=True)
+        self.hidden2tag = nn.Linear(hidden_dim, 1)
+        self.hidden = self.init_hidden()
+
+    def init_hidden(self):
```

Comment:
All these `2`s seem to be hardcodings. Take batch size as input and replace that in the code.